### ISSUE TITLE
Parsing works only when new rotated file and only once

### ISSUE THREAD

**Issue by CZAirwolf** [dalibo/pgbadger#258]
Date of the last ok update: 6.10.2015 (3330d885585b2d523b6f749fa9f741bb1f365c00)
Date of the update when parsing failed: 15.12.2015 (a60363a0b05d9732502705147d635856eff8afa0)
Current version: 25d5be6fe05b0ee5014205d64d3c561936641263

Issue: Incremental parsing works only for newly created rotated log and with "LAST_PARSED" dont work until next rotation. Rotation is around 6:55 AM. There is 6:55-7:xx analyze (xx depends on end of parsing), the other time lines are missing from both logs.

OS: Debian Wheezy, Jessie

ls -al:
-rw-r-----  1 postgres adm      1165256387 Jan  7 15:51 postgresql-9.3-main.log
-rw-r-----  1 postgres adm      6687103102 Jan  7 06:55 postgresql-9.3-main.log.1

crontab:
```
0 * * * * /usr/local/bin/pgbadger --prefix '\%t [\%p]: [\%l-1] user=\%u,db=\%d,client=\%h,appname=\%a ' -I -q /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/pgbadger/
```
```
root@SOMESRV /var/tmp/badger # cat /var/www/SOMEDIR/pgbadger/LAST_PARSED 
2016-01-07 07:00:31	15813345	2016-01-07 07:00:31 CET [16753]: [71-1] user=SOMEUSER,db=SOMEDB,client=SOMEIP,appname=[unknown] STATEMENT:  SOMEQUERY
```
```
root@SOMESRV /var/tmp/badger # egrep -n '16753.*71-1' /var/log/postgresql/postgresql-9.3-main.log
112931:2016-01-07 07:00:31 CET [16753]: [71-1] user=SOMEUSER,db=SOMEDB,client=SOMEIP,appname=[unknown] STATEMENT:  SOMEQUERY
```
```
root@SOMESRV /var/log/postgresql # /usr/local/bin/pgbadger --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -vvv /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/pgbadger/
DEBUG: Autodetected log format 'stderr' from /var/log/postgresql/postgresql-9.3-main.log
DEBUG: Output 'html' reports will be written to out.html
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log
DEBUG: the log statistics gathering took: 1 wallclock secs ( 0.06 usr +  0.14 sys =  0.20 CPU)
WARNING: can't read file /var/www/SOMEDIR/pgbadger//LAST_PARSED.tmp, No such file or directory
HINT: maybe there's no new entries in your log since last run.
LOG: Ok, generating global index to access incremental reports...
DEBUG: building reports took:  1 wallclock secs ( 0.07 usr +  0.05 sys =  0.12 CPU)
DEBUG: the total execution time took:  2 wallclock secs ( 0.13 usr +  0.19 sys =  0.32 CPU)
```
Maybe problem with line number in "LAST_PARSED"? Its higher than number in current log, lower than previous log, and still lower than previous+current line number.
```
root@SOMESRV /var/log/postgresql # wc -l postgresql-9.3-main.log
8515752 postgresql-9.3-main.log
root@SOMESRV /var/log/postgresql # wc -l postgresql-9.3-main.log.1
40195091 postgresql-9.3-main.log.1
```

**Comment by darold** [comment]
Hi,

This is not the last line number parsed that is stored in the LAST_PARSED file but the last byte offset. pgBadger will simply perform a seek() to that offset and look if next line has a timestamp before or after the history time stored in LAST_PARSED file too.

None of the commit you've pointed above are related to this code part, I will investigate further later about this issue this weekend.

Regards,

**Comment by CZAirwolf** [comment]
I restored pgbadger from backup - version (3330d885585b2d523b6f749fa9f741bb1f365c00) and parsing is ok. So there can be some change between 6.10. and 15.12. what affected detecting LAST_PARSED line. I can make some tests with different versions if you send me instruction how to obtain them from git or send me more debugging current pgbadger.

**Comment by darold** [comment]
After some different tries I'm not able to reproduce the issue, your help to find the responsible commit.

You  can get all commit into a file using:

    git log > list_commit.txt

then to revert to a commit:

   git checkout <commit_number>

Regards,

**Comment by darold** [comment]
Hi,

Please dowload latest code from GitHub and give it a try using the debug output. I have more debug information about check log parsing decision. So running your command:

    /usr/local/bin/pgbadger --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -vvv \
                  /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log \
                  -O /var/www/SOMEDIR/pgbadger/

will give us more information on why pgBadger decide to not parse your log file.

Please post here the debug output.

Best regards,

**Comment by CZAirwolf** [comment]
Hi,

i tested problem with different git versions, but still didn't ended (files are huge and it cost a lot of time). But for you request about new version, there is current test:
1] last parse was around 21:xx yesterday, rotation today 6:55, started test around 9:30, version 72e2467 - parsing OK
2] last parse today 11:34, started test around 12:15, version 25d5be6 - parsing failed
3] last parse today 11:34, started test after 2], version 72e2467 - parsing failed

ad 1]
```
root@SOMESRV# /usr/local/bin/pgbadger -j 2 --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -v /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/www/pgbadger/
DEBUG: Autodetected log format 'stderr' from /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: Output 'html' reports will be written to out.html
DEBUG: Starting progressbar writer process
DEBUG: Starting to parse log file: /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: Starting reading file /var/log/postgresql/postgresql-9.3-main.log.1...
DEBUG: Start parsing at offset 0 of file /var/log/postgresql/postgresql-9.3-main.log.1 to 32091758796
DEBUG: Starting to parse log file: /var/log/postgresql/postgresql-9.3-main.log
DEBUG: Starting reading file /var/log/postgresql/postgresql-9.3-main.log...
DEBUG: Start parsing at offset 0 of file /var/log/postgresql/postgresql-9.3-main.log to 2192738495
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16905 at line 128511711, events: 3841
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16534 at line 12851425
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 12857991
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16534 at line 128691661, events: 3844
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16905 at line 128760756, events: 3844
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 12877853
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 128822522, events: 3846
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16905 at line 12882452
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 12883426
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 12883445
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16905 at line 12883669
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 12884392
DEBUG: Unrecognized line: LOG: skipping analyze of "sold_skipasses_axess_history" --- lock not available at line 12884677
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16743 at line 12884680
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16905 at line 12884695
DEBUG: Unrecognized line: LOG: skipping analyze of "structure_nodes2permissions" --- lock not available at line 12884731
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 16905 at line 12887517
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 29869 at line 153525928, events: 4547
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 30521 at line 15353948
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 30242 at line 15355810
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 29869 at line 15359809
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 29869 at line 153602267, events: 4548
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 30521 at line 15360712
DEBUG: Unrecognized line: LOG: sending cancel to blocking autovacuum PID 30521 at line 153726962, events: 4551
DEBUG: Unrecognized line: LOG: skipping analyze of "discount_codes" --- lock not available at line 15552611555
DEBUG: Unrecognized line: LOG: skipping analyze of "discount_codes" --- lock not available at line 15582370555
[========================>] Parsed 34284473417 bytes of 34284473417 (100.00%), queries: 21672633, events: 4587
DEBUG: the log statistics gathering took:8387 wallclock secs ( 0.41 usr  0.34 sys + 13148.78 cusr 72.40 csys = 13221.93 CPU)
LOG: Ok, generating HTML daily report into /var/www/SOMEDIR/www/pgbadger//2016/01/13/...
LOG: Ok, generating HTML daily report into /var/www/SOMEDIR/www/pgbadger//2016/01/14/...
LOG: Ok, generating HTML weekly report into /var/www/SOMEDIR/www/pgbadger//2016/week-03/...
LOG: Ok, generating global index to access incremental reports...
DEBUG: building reports took: 395 wallclock secs (355.21 usr +  3.52 sys = 358.73 CPU)
DEBUG: the total execution time took: 8782 wallclock secs (355.62 usr  3.86 sys + 13148.78 cusr 72.40 csys = 13580.66 CPU)
```
ad 2]
```
root@SOMESRV# /usr/local/bin/pgbadger -j 2 --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -v /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/www/pgbadger/
DEBUG: Autodetected log format 'stderr' from /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: Output 'html' reports will be written to out.html
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log
DEBUG: the log statistics gathering took: 0 wallclock secs ( 0.02 usr +  0.01 sys =  0.03 CPU)
WARNING: can't read file /var/www/SOMEDIR/www/pgbadger//LAST_PARSED.tmp, No such file or directory
HINT: maybe there's no new entries in your log since last run.
LOG: Ok, generating global index to access incremental reports...
DEBUG: building reports took:  0 wallclock secs ( 0.01 usr +  0.01 sys =  0.02 CPU)
DEBUG: the total execution time took:  0 wallclock secs ( 0.03 usr +  0.02 sys =  0.05 CPU)
```
ad 3]
```
root@SOMESRV# /usr/local/bin/pgbadger -j 2 --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -v /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/www/pgbadger/
DEBUG: Autodetected log format 'stderr' from /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: Output 'html' reports will be written to out.html
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log.1, timestamp 2016-01-13 09:27:53 read at offset 4907682855 is lower than saved timestamp: 2016-01-14 10:50:34
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log, timestamp 2016-01-14 10:49:41 read at offset 4907682855 is lower than saved timestamp: 2016-01-14 10:50:34
DEBUG: the log statistics gathering took: 0 wallclock secs ( 0.01 usr +  0.01 sys =  0.02 CPU)
WARNING: can't read file /var/www/SOMEDIR/www/pgbadger//LAST_PARSED.tmp, No such file or directory
HINT: maybe there's no new entries in your log since last run.
LOG: Ok, generating global index to access incremental reports...
DEBUG: building reports took:  0 wallclock secs ( 0.01 usr +  0.02 sys =  0.03 CPU)
DEBUG: the total execution time took:  0 wallclock secs ( 0.02 usr +  0.03 sys =  0.05 CPU)
```

In case 1], pgbadger parsed over 34284473417 bytes, because lines was added after read/parse started (~2 hour run). Maybe there lies the problem. And If this is expected (aka continuous parsing), it make small (solvable :-)) problem with regular crons, because parsing may not be finished before next cronrun.

**Comment by darold** [comment]
The first parsing reported ( 1] ) seems to not see any history file because the full content of both log files is parsed. The second says that there's no new entry in the log file but additional debug information are not present du to the commit change.

The third parsing says that both files have already been parsed. For file /var/log/postgresql/postgresql-9.3-main.log.1 it seems ok because timestamp at starting offset in the log file is 2016-01-13 09:27:53 which is lower than the saved timestamp 2016-01-14 10:50:34 of the last pgBader run.

The report on the second file /var/log/postgresql/postgresql-9.3-main.log is the same, timestamp at starting offset in the log file is 2016-01-14 10:49:41 which is also lower than the saved timestamp 2016-01-14 10:50:34.

Everything works as expected, I don't see any bug on pgbadger in these behaviors.

You point me to cron jobs overlapping and looking at your cron job it is really possible to have an overlap with a run each hour, huge log files and no multiprocess. Last commit add the creation of a pid file to prevent two pgbadger process to be executed at the same time. Please install latest development code and let's see if you still have the issue. Overlapping attempt will be traces in system log file.

Let me know

**Comment by CZAirwolf** [comment]
Hi, i will test last code. But you probably misunderstood my comment. I will make it cleaner: 
Point 1] - made offset in LAST_PARSED calculated from postgresql-9.3-main.log
             - file postgresql-9.3-main.log was modified in parsing process by adding more lines by postgres
Point 2+3] - file postgresql-9.3-main.log.1 isn't parsed, because LAST_PARSED is pointing to postgresql-9.3.main.log - thats OK
Point 2+3] - file postgresql-9.3-main.log isn't parsed, because offset in LAST_PARSED <> offset in postgresql-9.3-main.log - there is problem, why is offset different?

Now, current pgbadger version (manual run):

1] cat /var/www/SOMEDIR/www/pgbadger/LAST_PARSED 
```
2016-01-14 10:50:34	4907682855	2016-01-14 10:50:34 CET [3517]: [125-1] user=tmr_crm,db=tmr,client=127.0.0.1,appname=[unknown] LOG:  duration: 47.611 ms  statement: SOMESQL
```
2]This line is now in postgresql-9.3-main.log.1 (after rotation at morning)
egrep -b -A3 -B3 "\[3517\]: \[125-1\]" postgresql-9.3-main.log.1
```
4925601395-	            SOMESQLXX
4925601439-	            
4925601453-2016-01-14 10:50:34 CET [3557]: [57-1] user=tmr_crm,db=tmr,client=127.0.0.1,appname=[unknown] LOG:  duration: 0.353 ms  statement: SOMESQLXX
4925601797:2016-01-14 10:50:34 CET [3517]: [125-1] user=tmr_crm,db=tmr,client=127.0.0.1,appname=[unknown] LOG:  duration: 47.611 ms  statement: SOMESQL
4925602050-2016-01-14 10:50:34 CET [3549]: [56-1] user=tmr_crm,db=tmr,client=127.0.0.1,appname=[unknown] LOG:  duration: 5.588 ms  statement: 
4925602182-	            SOMESQLXX
4925602250-	            SOMESQLXX  
```
Egrep with -b is saying, offset is 4925601797. LAST_PARSED has 4907682855 - so DEBUG is right, its lower...It looks as log was modified by postgres (or rsyslog) in (+/-) SAME time as calculating offset for LAST_PARSED and added some lines before LAST_PARSED line --> offset is calculated miscorrectly (+parsing missed those added lines?).

3] ls -al /var/log/postgresql
```
-rw-r-----  1 postgres postgres  2102543867 Jan 15 09:28 postgresql-9.3-main.log
-rw-r-----  1 postgres postgres 27560222753 Jan 15 07:01 postgresql-9.3-main.log.1
```
4] pgbadger started
```
root@SRV# /usr/local/bin/pgbadger -j 2 --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -v /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/www/pgbadger/
DEBUG: Autodetected log format 'stderr' from /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: Output 'html' reports will be written to out.html
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log.1, timestamp 2016-01-14 10:49:41 read at offset 4907682855 is lower than saved timestamp: 2016-01-14 10:50:34
DEBUG: Starting progressbar writer process
DEBUG: Starting to parse log file: /var/log/postgresql/postgresql-9.3-main.log
DEBUG: Starting reading file /var/log/postgresql/postgresql-9.3-main.log...
DEBUG: Start parsing at offset 0 of file /var/log/postgresql/postgresql-9.3-main.log to 2114613681
[========================>] Parsed 2114599928 bytes of 2114599928 (100.00%), queries: 13280704, events: 2981
DEBUG: the log statistics gathering took:4699 wallclock secs ( 0.32 usr  0.28 sys + 4461.04 cusr 17.90 csys = 4479.54 CPU)
LOG: Ok, generating HTML daily report into /var/www/SOMEDIR/www/pgbadger//2016/01/15/...
LOG: Ok, generating HTML weekly report into /var/www/SOMEDIR/www/pgbadger//2016/week-03/...
LOG: Ok, generating global index to access incremental reports...
DEBUG: building reports took: 428 wallclock secs (407.83 usr +  3.43 sys = 411.26 CPU)
DEBUG: the total execution time took: 5127 wallclock secs (408.15 usr  3.71 sys + 4461.04 cusr 17.90 csys = 4890.80 CPU)
```

**Comment by darold** [comment]
Ok, thanks for the precision, it's clear now.

> Egrep with -b is saying, offset is 4925601797. LAST_PARSED has 4907682855 - so DEBUG is right,
> its lower...It looks as log was modified by postgres (or rsyslog) in (+/-) SAME time as calculating offset
> for LAST_PARSED and added some lines before LAST_PARSED line --> offset is calculated
> miscorrectly (+parsing missed those added lines?).

No this is not a problem of file modified  or not after the start or end of pgbadger. The offset is calculated as follow: `$current_offset += length($line);`. I think this is the problem, it doesn't take care of unicode and return the length in ascii character. This the only reason I can see for this offset difference. Last commit 8775053 might solves this problem. There was also a case that can produce this difference, this is when excluded line regex was set at command line, but in your example this is not the case. Whatever, the commit fix that too.

Let me know.

**Comment by darold** [comment]
Humm, no this is a silly deduction I've reverted the last commit. Now commit might fix the issue 2ed0ea8. The length was taken after removing CR (\r) from the file :-(

**Comment by CZAirwolf** [comment]
(2ed0ea8):
```
root@SOMESRV# /usr/local/bin/pgbadger --prefix '%t [%p]: [%l-1] user=%u,db=%d,client=%h,appname=%a ' -I -v /var/log/postgresql/postgresql-9.3-main.log.1 /var/log/postgresql/postgresql-9.3-main.log -O /var/www/SOMEDIR/www/pgbadger/
DEBUG: Autodetected log format 'stderr' from /var/log/postgresql/postgresql-9.3-main.log.1
DEBUG: Output 'html' reports will be written to out.html
DEBUG: this file has already been parsed: /var/log/postgresql/postgresql-9.3-main.log.1, timestamp 2016-01-14 14:09:10 read at offset 10926931996 is lower than saved timestamp: 2016-01-15 16:42:39
DEBUG: Starting to parse log file: /var/log/postgresql/postgresql-9.3-main.log
DEBUG: This might not be parsed: /var/log/postgresql/postgresql-9.3-main.log, timestamp 2016-01-15 07:01:44 read at offset 0 is lower than saved timestamp: 2016-01-15 16:42:39
DEBUG: Starting reading file /var/log/postgresql/postgresql-9.3-main.log...
DEBUG: Start parsing at offset 10926931996 of file /var/log/postgresql/postgresql-9.3-main.log to 11217671920
[========================>] Parsed 11363637112 bytes of 11217671920 (101.30%), queries: 1386266, events: 71
DEBUG: the log statistics gathering took:449 wallclock secs (443.64 usr +  1.63 sys = 445.27 CPU)
```
It looks promising! 
Found 2 things:
1] pidfile (even without -j) isn't deleted when parser ended
```
DEBUG: building reports took: 508 wallclock secs (500.39 usr +  3.91 sys = 504.30 CPU)
DEBUG: the total execution time took: 781 wallclock secs (770.90 usr +  4.76 sys = 775.66 CPU)
# ls -al /tmp
-rw-r--r--  1 root      root         5 Jan 15 17:42 pgbadger.pid
``
2] Logs are in current time (UTC+1). Graph lines shows current time (well, with GMT:-)) - thats ok. But X axis shows UTC+2? Example:
```
UTC time: 16:00:00
Graph line: On Fri, 15 Jan 2016 17:00:00 GMT (==UTC+1)
X axis: Fri 18:00
```

**Comment by darold** [comment]
Commit 0b5a3ab and 4c19a60 solve the pidfile issue. I will look to the timezone issue.

**Comment by darold** [comment]
Well, the problem with the timezone issue is that the Javascript library use the timezone of the browser where the script is executed. If you change the timezone on you computer you will see that the time shown in the graphs change following the timezone. If your  timezone is not the same as the one on your server, the only thing I can do is to add a command line option to change the time internally into pgbadger that will apply the given timezone to all timestamp  in the javascript dataset.

Last commit 518fe7b add this feature with option `--timezone`. For example:

    perl pgbadger postgresql.log --timezone='+01'

Best regards,

**Comment by darold** [comment]
I have just published a new release (v7.3) to fix this major bug. Thanks again for your help to find the source of the issue.
